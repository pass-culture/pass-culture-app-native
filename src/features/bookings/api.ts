import { useQuery } from 'react-query'

import { api } from 'api/api'
import { BookingsResponse } from 'api/gen'

export function useBookings() {
  return useQuery<BookingsResponse>('bookings', () => api.getnativev1bookings())
}
