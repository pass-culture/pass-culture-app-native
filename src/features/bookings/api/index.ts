export { useArchiveBookingMutation } from './useArchiveBookingMutation'
export { useBookings } from './useBookings'
export { useOngoingOrEndedBooking } from './useOngoingOrEndedBooking'
export { useEndedBookingFromOfferId } from './useEndedBookingFromOfferId'
export { useCancelBookingMutation } from './useCancelBookingMutation'
