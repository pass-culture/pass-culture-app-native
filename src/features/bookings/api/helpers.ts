import { t } from '@lingui/macro'

import { isApiError } from 'api/helpers'

export function extractApiErrorMessage(error: unknown) {
  let message = t`Une erreur est survenue`
  if (isApiError(error)) {
    const { content } = error as { content: { code: string; message: string } }
    if (content && content.code && content.message) {
      message = content.message
    }
  }

  return message
}
