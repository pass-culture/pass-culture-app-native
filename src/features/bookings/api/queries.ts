import { useCallback } from 'react'
import { useQuery, UseQueryResult } from 'react-query'

import { api } from 'api/api'
import { BookingReponse, BookingsResponse } from 'api/gen'
import { useNetwork } from 'libs/network/useNetwork'
import { QueryKeys } from 'libs/queryKeys'

export function useBookings(options = {}) {
  const { isConnected } = useNetwork()

  return useQuery<BookingsResponse>(QueryKeys.BOOKINGS, () => api.getnativev1bookings(), {
    enabled: isConnected,
    ...options,
  })
}

export function useOngoingOrEndedBooking(id: number): UseQueryResult<BookingReponse | null> {
  const select = useCallback(
    (bookings) => {
      if (!bookings) {
        return null
      }
      const onGoingBooking = bookings.ongoing_bookings.find(
        (item: BookingReponse) => item.id === id
      )
      const endedBooking = bookings.ended_bookings.find((item: BookingReponse) => item.id === id)

      const selected = onGoingBooking || endedBooking
      if (!selected) {
        return null
      }
      return selected
    },
    [id]
  )

  return useBookings({
    select,
  }) as UseQueryResult<BookingReponse | null>
}
