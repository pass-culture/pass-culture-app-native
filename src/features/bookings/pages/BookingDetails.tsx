import { t } from '@lingui/macro'
import { useNavigation, useRoute } from '@react-navigation/native'
import React from 'react'
import { ScrollView, StyleSheet, useWindowDimensions } from 'react-native'
import { useQueryClient } from 'react-query'
import styled from 'styled-components/native'

import { useAppSettings } from 'features/auth/settings'
import { useOngoingOrEndedBooking } from 'features/bookings/api/queries'
import { ArchiveBookingModal } from 'features/bookings/components/ArchiveBookingModal'
import { BookingDetailsCancelButton } from 'features/bookings/components/BookingDetailsCancelButton'
import { BookingDetailsHeader } from 'features/bookings/components/BookingDetailsHeader'
import { BookingDetailsTicketContent } from 'features/bookings/components/BookingDetailsTicketContent'
import { BookingPropertiesSection } from 'features/bookings/components/BookingPropertiesSection'
import { CancelBookingModal } from 'features/bookings/components/CancelBookingModal'
import { ThreeShapesTicket } from 'features/bookings/components/ThreeShapesTicket'
import { BookingProperties, getBookingProperties } from 'features/bookings/helpers'
import { BookingNotFound } from 'features/bookings/pages/BookingNotFound'
import { UseNavigationType, UseRouteType } from 'features/navigation/RootNavigator'
import { mergeOfferData } from 'features/offer/atoms/OfferTile'
import { analytics, isCloseToBottom } from 'libs/analytics'
import useFunctionOnce from 'libs/hooks/useFunctionOnce'
import { SeeItineraryButton } from 'libs/itinerary/components/SeeItineraryButton'
import { getGoogleMapsItineraryUrl } from 'libs/itinerary/openGoogleMapsItinerary'
import useOpenItinerary from 'libs/itinerary/useOpenItinerary'
import { ScreenError } from 'libs/monitoring'
import { QueryKeys } from 'libs/queryKeys'
import { GLOBAL_STALE_TIME } from 'libs/react-query/queryClient'
import { useSubcategoriesMapping } from 'libs/subcategories'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { useHeaderTransition } from 'ui/components/headers/animationHelpers'
import { HeroHeader } from 'ui/components/hero/HeroHeader'
import { blurImageHeight, heroMarginTop } from 'ui/components/hero/useHeroDimensions'
import { LoadingPage } from 'ui/components/LoadingPage'
import { useModal } from 'ui/components/modals/useModal'
import { Separator } from 'ui/components/Separator'
import { getSpacing, Spacer, Typo } from 'ui/theme'
import { Helmet } from 'ui/web/global/Helmet'
import { A } from 'ui/web/link/A'
import { Link } from 'ui/web/link/Link'
const getOfferRules = (
  properties: BookingProperties,
  activationCodeFeatureEnabled?: boolean
): string => {
  const { hasActivationCode, isDigital, isPhysical, isEvent } = properties
  if (hasActivationCode && activationCodeFeatureEnabled)
    return t`Ce code est ta preuve d’achat, il te permet d’accéder à ton offre\u00a0! N’oublie pas que tu n’as pas le droit de le revendre ou le céder.`
  if (isDigital)
    return t`Ce code à 6 caractères est ta preuve d’achat\u00a0! N’oublie pas que tu n’as pas le droit de le revendre ou le céder.`
  if (isPhysical || isEvent)
    return t`Tu dois présenter ta carte d’identité et ce code de 6 caractères pour profiter de ta réservation\u00a0! N’oublie pas que tu n’as pas le droit de le revendre ou le céder.`
  return ''
}

const scrollIndicatorInsets = { right: 1 }

export function BookingDetails() {
  const windowHeight = useWindowDimensions().height - blurImageHeight
  const { params } = useRoute<UseRouteType<'BookingDetails'>>()
  const { navigate } = useNavigation<UseNavigationType>()
  const { data: booking, isLoading, isError, error } = useOngoingOrEndedBooking(params.id)
  const queryClient = useQueryClient()
  const { visible: cancelModalVisible, showModal: showCancelModal, hideModal } = useModal(false)
  const {
    visible: archiveModalVisible,
    showModal: showArchiveModal,
    hideModal: hideArchiveModal,
  } = useModal(false)

  const { data: appSettings } = useAppSettings()
  const mapping = useSubcategoriesMapping()

  const { venue, id: offerId } = booking?.stock.offer || {}
  const { address, postalCode, city } = venue || {}
  const { canOpenItinerary, openItinerary } = useOpenItinerary(
    `${address} ${postalCode} ${city}`,
    async () => void (offerId && analytics.logConsultItinerary({ offerId, from: 'bookingdetails' }))
  )

  const logConsultWholeBooking = useFunctionOnce(
    () => offerId && analytics.logBookingDetailsScrolledToBottom(offerId)
  )

  const { headerTransition, onScroll } = useHeaderTransition({
    listener: ({ nativeEvent }) => {
      if (isCloseToBottom(nativeEvent)) logConsultWholeBooking()
    },
  })

  if (isLoading && !booking) {
    return <LoadingPage />
  } else if (!isLoading && !booking) {
    throw new ScreenError(`Booking #${params.id} not found`, BookingNotFound)
  } else if (isError) {
    throw error
  } else if (!booking) {
    // dead code to satisfy typescript Web compilation
    return null
  }

  const { offer } = booking.stock
  const properties = getBookingProperties(booking, mapping[offer.subcategoryId].isEvent)
  const shouldDisplayItineraryButton =
    canOpenItinerary && (properties.isEvent || (properties.isPhysical && !properties.isDigital))
  const activationCodeFeatureEnabled = appSettings && appSettings.autoActivateDigitalBookings

  const offerRules = getOfferRules(properties, activationCodeFeatureEnabled)

  const cancelBooking = () => {
    showCancelModal()
    analytics.logCancelBooking(offer.id)
  }

  const navigateToOffer = () => {
    queryClient.setQueryData(
      [QueryKeys.OFFER, offer.id],
      mergeOfferData({
        ...offer,
        categoryId: mapping[offer.subcategoryId].categoryId,
        thumbUrl: offer.image?.url,
        name: offer.name,
        offerId: offer.id,
      }),
      {
        // Make sure the data is stale, so that it is considered as a placeholder
        updatedAt: Date.now() - (GLOBAL_STALE_TIME + 1),
      }
    )
    analytics.logConsultOffer({ offerId: offer.id, from: 'bookings' })
    navigate('Offer', { id: offer.id, from: 'bookingdetails' })
  }

  const helmetTitle = `${t`Ma réservation pour`} ${booking.stock.offer.name} | pass Culture`
  return (
    <Container>
      <Helmet title={helmetTitle} />
      <ScrollView
        onScroll={onScroll}
        scrollEventThrottle={20}
        scrollIndicatorInsets={scrollIndicatorInsets}
        onContentSizeChange={(_w: number, h: number) => {
          if (h <= windowHeight) {
            logConsultWholeBooking()
          }
        }}
        testID="BookingDetailsScrollView"
        bounces={false}>
        <HeroHeader type="offer" imageHeight={blurImageHeight} imageUrl={offer.image?.url}>
          <Spacer.Column numberOfSpaces={heroMarginTop} />
          <ThreeShapesTicket>
            <BookingDetailsTicketContent
              booking={booking}
              activationCodeFeatureEnabled={activationCodeFeatureEnabled}
            />
          </ThreeShapesTicket>
        </HeroHeader>

        <ViewWithPadding>
          <Spacer.Column numberOfSpaces={4} />
          <OfferRules>{offerRules}</OfferRules>
          <Spacer.Column numberOfSpaces={8} />
          <BookingPropertiesSection booking={booking} />
          {!!shouldDisplayItineraryButton && (
            <React.Fragment>
              <Spacer.Column numberOfSpaces={4} />
              <Separator />
              <Spacer.Column numberOfSpaces={4} />
              <A href={address ? getGoogleMapsItineraryUrl(address) : undefined}>
                <SeeItineraryButton openItinerary={openItinerary} />
              </A>
            </React.Fragment>
          )}
          {!!offer.withdrawalDetails && (
            <React.Fragment>
              <Spacer.Column numberOfSpaces={8} />
              <Typo.Title4>{t`Modalités de retrait`}</Typo.Title4>
              <Spacer.Column numberOfSpaces={4} />
              <Typo.Body testID="withdrawalDetails">{offer.withdrawalDetails}</Typo.Body>
            </React.Fragment>
          )}
          <Spacer.Column numberOfSpaces={8} />
          <Link
            to={{ screen: 'Offer', params: { id: offer.id, from: 'bookingdetails' } }}
            style={styles.link}
            accessible={false}>
            <ButtonPrimary
              testID="Voir le détail de l’offre"
              wording={t`Voir le détail de l’offre`}
              onPress={navigateToOffer}
              fullWidth
            />
          </Link>
          <Spacer.Column numberOfSpaces={4} />
          <BookingDetailsCancelButton
            booking={booking}
            activationCodeFeatureEnabled={activationCodeFeatureEnabled}
            onCancel={cancelBooking}
            onTerminate={showArchiveModal}
            fullWidth
          />
        </ViewWithPadding>
        <Spacer.Column numberOfSpaces={5} />
      </ScrollView>

      <CancelBookingModal visible={cancelModalVisible} dismissModal={hideModal} booking={booking} />
      <ArchiveBookingModal
        visible={archiveModalVisible}
        bookingId={booking.id}
        bookingTitle={offer.name}
        onDismiss={hideArchiveModal}
      />
      <BookingDetailsHeader headerTransition={headerTransition} title={offer.name} />
    </Container>
  )
}

const Container = styled.View(({ theme }) => ({
  flex: 1,
  backgroundColor: theme.colors.white,
}))

const OfferRules = styled(Typo.Caption)(({ theme }) => ({
  color: theme.colors.greyDark,
  textAlign: 'center',
}))

const ViewWithPadding = styled.View({
  paddingHorizontal: getSpacing(5),
})

const styles = StyleSheet.create({
  link: {
    flexDirection: 'column',
    display: 'flex',
  },
})
