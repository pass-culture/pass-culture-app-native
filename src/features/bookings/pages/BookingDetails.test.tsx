import React from 'react'

import { navigate, useRoute } from '__mocks__/@react-navigation/native'
import { CategoryNameEnum, CategoryType } from 'api/gen'
import * as Queries from 'features/bookings/api/queries'
import * as Helpers from 'features/bookings/helpers'
import * as NavigationHelpers from 'features/navigation/helpers'
import { analytics } from 'libs/analytics'
import * as OpenItinerary from 'libs/itinerary/useOpenItinerary'
import { reactQueryProviderHOC } from 'tests/reactQueryProviderHOC'
import { act, fireEvent, render } from 'tests/utils'

import { bookingsSnap } from '../api/bookingsSnap'
import { Booking } from '../components/types'

import { BookingDetails } from './BookingDetails'

jest.mock('features/home/api', () => ({
  useUserProfileInfo: jest.fn(() => ({ data: undefined })),
}))
jest.mock('libs/itinerary/useItinerary', () => ({
  useItinerary: jest.fn(() => ({ availableApps: ['waze'], navigateTo: jest.fn() })),
}))

describe('BookingDetails', () => {
  afterEach(jest.restoreAllMocks)

  beforeAll(() => {
    useRoute.mockImplementation(() => ({
      params: {
        id: 456,
        shouldFetchAll: true,
      },
    }))
  })

  it('should call useOngoingBooking with the right parameters', () => {
    const useOngoingBooking = jest.spyOn(Queries, 'useOngoingBooking')

    const booking = bookingsSnap.ongoing_bookings[0]
    renderBookingDetails(booking)
    expect(useOngoingBooking).toBeCalledWith(456, true)
  })

  it('should render correctly', async () => {
    const booking = bookingsSnap.ongoing_bookings[0]
    const { getByTestId, toJSON } = renderBookingDetails(booking)

    await act(async () => {
      getByTestId('three-shapes-ticket').props.onLayout({ nativeEvent: { layout: { width: 150 } } })
    })
    expect(toJSON()).toMatchSnapshot()
  })

  describe('<DetailedBookingTicket />', () => {
    it('should display booking token', async () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      const { getByText, getByTestId } = renderBookingDetails(booking)
      await act(async () => {
        getByTestId('three-shapes-ticket').props.onLayout({
          nativeEvent: { layout: { width: 150 } },
        })
      })

      getByText('352UW4')
    })

    it('should display offer link button if offer is digital and open url on press', async () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      const openExternalUrl = jest
        .spyOn(NavigationHelpers, 'openExternalUrl')
        .mockImplementation(jest.fn())
      booking.stock.offer.isDigital = true
      booking.stock.offer.url = 'http://example.com'

      const { getByText, getByTestId } = renderBookingDetails(booking)
      await act(async () => {
        getByTestId('three-shapes-ticket').props.onLayout({
          nativeEvent: { layout: { width: 150 } },
        })
      })
      const offerButton = getByText("Accéder à l'offre")
      fireEvent.press(offerButton)

      expect(openExternalUrl).toHaveBeenCalledWith(booking.stock.offer.url)
      expect(analytics.logAccessExternalOffer).toHaveBeenCalledWith(booking.stock.offer.id)
    })

    it('should display booking qr code if offer is physical', async () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      booking.stock.offer.isDigital = false
      const { getByTestId } = renderBookingDetails(booking)
      await act(async () => {
        getByTestId('three-shapes-ticket').props.onLayout({
          nativeEvent: { layout: { width: 150 } },
        })
      })
      getByTestId('qr-code')
    })

    it('should display EAN code if offer is a book (digital or physical)', async () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      booking.stock.offer.category.name = CategoryNameEnum.LIVRE
      const { getByText, getByTestId } = renderBookingDetails(booking)
      await act(async () => {
        getByTestId('three-shapes-ticket').props.onLayout({
          nativeEvent: { layout: { width: 150 } },
        })
      })
      getByText('123456789')
    })
  })

  describe('Offer rules', () => {
    it('should display rules for a digital offer', () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      booking.stock.offer.isDigital = true

      const { getByText } = renderBookingDetails(booking)

      getByText(
        'Ce code à 6 caractères est ta preuve d’achat ! N’oublie pas que tu n’as pas le droit de le revendre ou le céder.'
      )
    })
    it.each([
      [
        'event',
        (booking: Booking) => (booking.stock.offer.category.categoryType = CategoryType.Event),
      ],
      [
        'physical',
        (booking: Booking) => (booking.stock.offer.category.categoryType = CategoryType.Thing),
      ],
    ])('should display rules for a %s & non-digital offer', (type, prepareBooking) => {
      const booking = { ...bookingsSnap.ongoing_bookings[0] }
      booking.stock.offer.isDigital = false
      prepareBooking(booking)

      const { getByText } = renderBookingDetails(booking)

      getByText(
        'Tu dois présenter ta carte d’identité et ce code de 6 caractères pour profiter de ta réservation ! N’oublie pas que tu n’as pas le droit de le revendre ou le céder.'
      )
    })
  })

  describe('withdrawalDetails', () => {
    it('should display withdrawal details', () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      booking.stock.offer.withdrawalDetails = 'Voici comment récupérer ton bien'
      const { getByText } = renderBookingDetails(booking)

      getByText('Modalités de retrait')
      getByText(booking.stock.offer.withdrawalDetails)
    })
    it('should not display withdrawal details', () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      booking.stock.offer.withdrawalDetails = undefined
      const { queryByTestId, queryByText } = renderBookingDetails(booking)

      const title = queryByText('Modalités de retrait')
      const withdrawalText = queryByTestId('withdrawalDetails')

      expect(title).toBeFalsy()
      expect(withdrawalText).toBeFalsy()
    })
  })
  it('should redirect to the Offer page and log event', () => {
    const booking = bookingsSnap.ongoing_bookings[0]
    const { getByText } = renderBookingDetails(booking)

    const text = getByText('Voir le détail de l’offre')
    fireEvent.press(text)

    const offerId = booking.stock.offer.id

    expect(navigate).toBeCalledWith('Offer', {
      id: offerId,
      shouldDisplayLoginModal: false,
      from: 'bookingdetails',
    })
    expect(analytics.logConsultOffer).toHaveBeenCalledWith({ offerId, from: 'bookings' })
  })

  describe('cancellation button', () => {
    it('should display button if offer is permanent', () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      booking.stock.offer.isPermanent = true
      const { getByTestId } = renderBookingDetails(booking)
      getByTestId('button-title-cancel')
    })

    it('should display button if confirmation date is not expired', () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      const date = new Date()
      date.setDate(date.getDate() + 1)
      booking.confirmationDate = date
      const { getByTestId } = renderBookingDetails(booking)
      getByTestId('button-title-cancel')
    })

    it('should not display button if confirmation date is expired', async () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      booking.stock.offer.isPermanent = false
      booking.confirmationDate = new Date('2020-03-15T23:01:37.925926')
      const { queryByTestId } = renderBookingDetails(booking)
      expect(queryByTestId('button-title-cancel')).toBeFalsy()
    })

    it('should log event "CancelBooking" when cancelling booking', () => {
      const booking = bookingsSnap.ongoing_bookings[0]
      const date = new Date()
      date.setDate(date.getDate() + 1)
      booking.confirmationDate = date
      const { getByTestId } = renderBookingDetails(booking)
      fireEvent.press(getByTestId('button-title-cancel'))

      expect(analytics.logCancelBooking).toHaveBeenCalledWith(booking.stock.offer.id)
    })
  })

  describe('Itinerary', () => {
    it.each([
      ['isEvent == true', { isEvent: true }],
      ['isPhysical == true', { isPhysical: true }],
    ])('should render the itinerary button when %s', (_testLabel, dataProvider) => {
      const openItinerary = jest.spyOn(OpenItinerary, 'default').mockReturnValue({
        openItinerary: jest.fn(),
        canOpenItinerary: true,
      })
      const getBookingProperties = jest
        .spyOn(Helpers, 'getBookingProperties')
        .mockReturnValue(dataProvider)

      const booking = bookingsSnap.ongoing_bookings[0]
      const { getByText } = renderBookingDetails(booking)
      getByText("Voir l'itinéraire")

      openItinerary.mockRestore()
      getBookingProperties.mockRestore()
    })
    it.each([
      ['canOpenItinerary == false', false, {}],
      [
        'canOpenItinerary == true && isEvent == false && isPhysical == false',
        true,
        { isEvent: false, isPhysical: false },
      ],
    ])(
      'should not render the itinerary button when %s',
      (_testLabel, canOpenItinerary, dataProvider) => {
        const openItinerary = jest.spyOn(OpenItinerary, 'default').mockReturnValue({
          openItinerary: jest.fn(),
          canOpenItinerary,
        })
        const getBookingProperties = jest
          .spyOn(Helpers, 'getBookingProperties')
          .mockReturnValue(dataProvider)

        const booking = bookingsSnap.ongoing_bookings[0]
        const { queryByText } = renderBookingDetails(booking)
        const button = queryByText("Voir l'itinéraire")
        expect(button).toBeFalsy()

        openItinerary.mockRestore()
        getBookingProperties.mockRestore()
      }
    )
  })
})

function renderBookingDetails(booking: Booking) {
  jest.spyOn(Queries, 'useOngoingBooking').mockReturnValue(booking)
  return render(reactQueryProviderHOC(<BookingDetails />))
}
