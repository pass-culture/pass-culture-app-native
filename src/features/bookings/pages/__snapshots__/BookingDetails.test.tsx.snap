// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingDetails should render correctly 1`] = `
Array [
  <RCTScrollView
    bounces={false}
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onScroll={[Function]}
    scrollEventThrottle={10}
    scrollIndicatorInsets={
      Object {
        "right": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "position": "absolute",
              },
            ]
          }
        >
          <View
            height={312}
            testID="BackgroundPlaceholder"
            width={751}
          >
            <Text>
              BackgroundPlaceholder-SVG-Mock
            </Text>
          </View>
          <View
            height={8}
            width={751}
          >
            <Text>
              undefined-SVG-Mock
            </Text>
          </View>
        </View>
        <View
          numberOfSpaces={18}
          style={
            Array [
              Object {
                "height": 72,
              },
            ]
          }
        />
        <View
          customWitdh={300}
          onLayout={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "column",
                "justifyContent": "center",
                "maxWidth": 300,
              },
              Object {
                "shadowColor": "#151515",
                "shadowOffset": Object {
                  "height": 2,
                  "width": 0,
                },
                "shadowOpacity": 0.1,
                "shadowRadius": 3,
              },
            ]
          }
          testID="three-shapes-ticket"
        >
          <View
            height={37.62376237623762}
            width={300}
          >
            <Text>
              undefined-SVG-Mock
            </Text>
          </View>
          <View
            color="#ffffff"
            customWitdh={150}
            style={
              Array [
                Object {
                  "backgroundColor": "#ffffff",
                  "marginTop": -1,
                  "width": 150,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "minHeight": 220,
                    "paddingHorizontal": 28,
                    "paddingVertical": 8,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#151515",
                      "fontFamily": "Montserrat-Bold",
                      "fontSize": 20,
                      "lineHeight": 24,
                      "paddingHorizontal": 4,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Avez-vous déjà vu ?
              </Text>
              <View
                numberOfSpaces={8}
                style={
                  Array [
                    Object {
                      "height": 32,
                    },
                  ]
                }
              />
              <Text
                style={
                  Array [
                    Object {
                      "color": "#eb0055",
                      "fontFamily": "Montserrat-Medium",
                      "fontSize": 18,
                      "lineHeight": 22,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                352UW4
              </Text>
              <View
                numberOfSpaces={2.5}
                style={
                  Array [
                    Object {
                      "height": 10,
                    },
                  ]
                }
              />
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#eb0055",
                    "borderRadius": 24,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "height": 40,
                    "justifyContent": "center",
                    "maxWidth": 500,
                    "opacity": 1,
                    "paddingBottom": 2,
                    "paddingLeft": 2,
                    "paddingRight": 2,
                    "paddingTop": 2,
                    "width": "100%",
                  }
                }
                testID="button-container"
              >
                <Text
                  adjustsFontSizeToFit={false}
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#ffffff",
                        "fontFamily": "Montserrat-Bold",
                        "fontSize": 15,
                        "lineHeight": 20,
                        "marginLeft": 5,
                        "maxWidth": 650,
                      },
                    ]
                  }
                  testID="button-title"
                  textColor="#ffffff"
                >
                  Accéder à l'offre
                </Text>
              </View>
              <View
                numberOfSpaces={9}
                style={
                  Array [
                    Object {
                      "height": 36,
                    },
                  ]
                }
              />
            </View>
          </View>
          <View
            height={37.62376237623762}
            width={300}
          >
            <Text>
              undefined-SVG-Mock
            </Text>
          </View>
        </View>
      </View>
      <View
        numberOfSpaces={4}
        style={
          Array [
            Object {
              "height": 16,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#c7c7cc",
              "fontFamily": "Montserrat-SemiBold",
              "fontSize": 12,
              "lineHeight": 16,
              "paddingHorizontal": 24,
              "textAlign": "center",
            },
          ]
        }
      >
        Ce code à 6 caractères est ta preuve d’achat ! N’oublie pas que tu n’as pas le droit de le revendre ou le céder.
      </Text>
      <View
        numberOfSpaces={8}
        style={
          Array [
            Object {
              "height": 32,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "paddingHorizontal": 20,
            },
          ]
        }
      >
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#151515",
                  "fontFamily": "Montserrat-Medium",
                  "fontSize": 18,
                  "lineHeight": 22,
                },
              ]
            }
          >
            Ma réservation
          </Text>
          <View
            numberOfSpaces={4.5}
            style={
              Array [
                Object {
                  "height": 18,
                },
              ]
            }
          />
          <View
            numberOfSpaces={5}
            style={
              Array [
                Object {
                  "height": 20,
                },
              ]
            }
          />
          <View
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="section-row-touchable"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  },
                  undefined,
                ]
              }
            >
              <View
                height={24}
                width={24}
              >
                <Text>
                  undefined-SVG-Mock
                </Text>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 1,
                      "marginHorizontal": 8,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#151515",
                        "fontFamily": "Montserrat-Regular",
                        "fontSize": 15,
                        "lineHeight": 20,
                      },
                    ]
                  }
                >
                  Le 15 mars 2021 à 20h00
                </Text>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "flex-end",
                    },
                  ]
                }
              />
            </View>
          </View>
          <View
            numberOfSpaces={5}
            style={
              Array [
                Object {
                  "height": 20,
                },
              ]
            }
          />
        </View>
        <View
          numberOfSpaces={4}
          style={
            Array [
              Object {
                "height": 16,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#f5f5f5",
                "height": 1,
                "width": "100%",
              },
            ]
          }
        />
        <View
          numberOfSpaces={4}
          style={
            Array [
              Object {
                "height": 16,
              },
            ]
          }
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "opacity": 1,
            }
          }
        >
          <View
            height={24}
            width={24}
          >
            <Text>
              undefined-SVG-Mock
            </Text>
          </View>
          <View
            numberOfSpaces={1}
            style={
              Array [
                Object {
                  "width": 4,
                },
              ]
            }
          />
          <Text
            style={
              Array [
                Object {
                  "color": "#151515",
                  "fontFamily": "Montserrat-Bold",
                  "fontSize": 15,
                  "lineHeight": 20,
                },
              ]
            }
          >
            Voir l'itinéraire
          </Text>
        </View>
      </View>
      <View
        numberOfSpaces={50}
        style={
          Array [
            Object {
              "height": 200,
            },
          ]
        }
      />
    </View>
  </RCTScrollView>,
  <View
    style={
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0)",
        "position": "absolute",
        "top": 0,
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <View
      customHeight={16}
      style={
        Array [
          Object {
            "height": 16,
          },
        ]
      }
    />
    <View
      numberOfSpaces={2}
      style={
        Array [
          Object {
            "height": 8,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingLeft": 12,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "aspectRatio": 1,
                "backgroundColor": "rgba(255, 255, 255, 1)",
                "borderColor": "rgba(245, 245, 245, 1)",
                "borderRadius": 40,
                "borderStyle": "solid",
                "borderWidth": 1,
                "justifyContent": "center",
                "overflow": "hidden",
                "transform": undefined,
                "width": 40,
              }
            }
            testID="headerIconRoundContainer"
          >
            <View
              size={32}
              style={
                Array [
                  Object {
                    "height": 32,
                    "width": 32,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "opacity": 1,
                    "position": "absolute",
                  }
                }
                testID="initial-icon-container"
              >
                <View
                  height={32}
                  testID="icon-back"
                  width={32}
                >
                  <Text>
                    icon-back-SVG-Mock
                  </Text>
                </View>
              </View>
              <View
                style={
                  Object {
                    "opacity": 0,
                    "position": "absolute",
                  }
                }
                testID="final-icon-container"
              >
                <View
                  height={32}
                  testID="final-icon-back"
                  width={32}
                >
                  <Text>
                    final-icon-back-SVG-Mock
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <Text
        numberOfLines={1}
        style={
          Object {
            "flexBasis": 0,
            "flexGrow": 5,
            "flexShrink": 1,
            "opacity": 0,
            "textAlign": "center",
          }
        }
      >
        <Text
          color="#ffffff"
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontFamily": "Montserrat-Regular",
                "fontSize": 15,
                "lineHeight": 20,
              },
            ]
          }
        >
          Avez-vous déjà vu ?
        </Text>
      </Text>
      <View
        numberOfSpaces={3}
        style={
          Array [
            Object {
              "width": 12,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
    </View>
    <View
      numberOfSpaces={2}
      style={
        Array [
          Object {
            "height": 8,
          },
        ]
      }
    />
  </View>,
]
`;
